### PR TITLE
(Generic Assay) Import Patient Level data

### LINKED ISSUE #8459: (Generic Assay) Support patient level and normal sample data
Based on the discussion: https://github.com/cBioPortal/cbioportal/issues/8450#issuecomment-804297016

We want to accomplish this by duplicating data to tumor sample level. (e.g. If we have data for `patient_1`, and `patient_1` has two samples (`sample_1` and `sample_2`), we need to find out that those two samples are related and duplicate data for both `sample_1` and `sample_2`).

- [x] Update validator
- [x] Change database scheme to introduce a new bool field `PATIENT_LEVEL` (e.g. PATIENT_LEVEL: `1` / `0`) into genetic_profile
- [x] Update importer to allow new type of data to be imported (if data is patient level or normal sample, adding data for all related tumor samples)
- [x] (Backend) Comparison endpoints: for patient data, return patients level calculation, not samples level calculation
- [x] (Frontend) Study view: select and filter by selected patients
- [x] Update `File-Formats.md`

### PR DESCRIPTION
**Pull request by dippindots**
Fix #8459 

Describe changes proposed in this pull request:
- Include database migration (new field in `genetic_profile`: `PATIENT_LEVEL`, default value is `false`)
- Data validator
- New data importer: `ImportGenericAssayPatientLevelData` and tests.
- Update logic in `ExpressionEnrichmentServiceImpl.java` to filter samples by unique patient.
- Update logic in `StudyViewFilterApplier.java` to filter samples by unique patient.

### CODE REVIEW

**Review by n1zea144 — CHANGES_REQUESTED** [review]
I'm not entirely sure why there is `patient_level` generic assay data.  It seems like when patient_level is true, data is propagated across all samples.  Is patient level data supported to make it more convenient to apply a measure across all samples that belong to the patient?

**Comment by dippindots** [comment]
@n1zea144 Thanks for your review, this was part of our discussion with @kalletlak and @jjgao. 
We thought about a few different ways to support `patient level` data. And Rest of them will require much more time to implement, so we decided to import `patient level` data as `sample level` data since we can reuse all existing implementation and just require a little bit changes to the front-end. https://github.com/cBioPortal/cbioportal/issues/8450#issuecomment-804297016

**Comment by n1zea144** [comment]
If patient level data will always be imported as sample level data, why not change the file format to accept sample ids instead of patient ids and get rid of the patient level boolean in the meta properties flag.  It would make the code/documentation less complicated? cc: @kalletlak @jjgao

File: docs/File-Formats.md [hunk 1]
```diff
@@ -1409,6 +1411,9 @@ values to the left.  When `value_sort_order` is `DESC` the x-axis will be in des
 ### Note on `generic_entity_meta_properties`
 All meta properties must be specified in the `generic_entity_meta_properties` field. Every meta property listed here must appear as a column header in the corresponding data file. It's highly recommend to add `NAME`, `DESCRIPTION` and an optional `URL` to get the best visualization on OncoPrint tab and Plots tab.
 
+### Note on `patient_level`
+Data will be considered as sample level data if `patient_level` is not specified or is `false`. And patients or samples need to be included in clinical data file. Please check the [Clinical Data](#clinical-data) section for more information.
```
[the remark this answers is not in the document]
Comment (dippindots) [reply docs/File-Formats.md:1415]: Changed to suggested text, UI will not be affected visually, just logic for statistics will change.

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGenericAssayPatientLevelData.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        Integer startFeatureCol = 1;
+        
+        // list the names of feature columns here
+        List<String> featureColNames = new ArrayList<String>();
+        featureColNames.add("ENTITY_STABLE_ID");
+
+        // add genericEntityProperties as feature colums
+        if (genericEntityProperties != null && genericEntityProperties.trim().length() != 0) {
+            String[] propertyNames = genericEntityProperties.trim().split(",");
+            featureColNames.addAll(Arrays.asList(propertyNames));
```
[the remark this answers is not in the document]
Comment (dippindots) [reply core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGenericAssayPatientLevelData.java:291]: No, `genericEntityProperties` is a string in meta files that defines additional columns for `Generic Assay`. `ENTITY_STABLE_ID` is a fixed column which will not show up in that string.

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGenericAssayPatientLevelData.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+            }
+
+            ProgressMonitor.setCurrentMessage(" --> total number of data lines:  " + (numLines-1));
+            
+            DaoGeneticProfileSamples.addGeneticProfileSamples(geneticProfileId, orderedSampleList);
+    
+            //Object to insert records in the generic 'genetic_alteration' table: 
+            DaoGeneticAlteration daoGeneticAlteration = DaoGeneticAlteration.getInstance();              
+            
+            int lenParts = parts.length;
```
[the remark this answers is not in the document]
Comment (dippindots) [reply core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGenericAssayPatientLevelData.java]: This is useless, removed.

File: docs/File-Formats.md [hunk 4]
```diff
@@ -1409,6 +1411,9 @@ values to the left.  When `value_sort_order` is `DESC` the x-axis will be in des
 ### Note on `generic_entity_meta_properties`
 All meta properties must be specified in the `generic_entity_meta_properties` field. Every meta property listed here must appear as a column header in the corresponding data file. It's highly recommend to add `NAME`, `DESCRIPTION` and an optional `URL` to get the best visualization on OncoPrint tab and Plots tab.
 
+### Note on `patient_level`
+Generic Assay data will be considered `sample_level` data if the `patient_level` property is missing or set to `false`. In addition, the patient or sample identifiers need to be include in the [Clinical Data](#clinical-data) file.
```
Comment (n1zea144) [inline docs/File-Formats.md:1415]: Minor typo s/include/included/

**Review by n1zea144 — CHANGES_REQUESTED** [review]

File: docs/File-Formats.md (hunk 4 above)
Comment (dippindots) [reply docs/File-Formats.md]: fixed, thanks!

**Review by n1zea144 — APPROVED** [review]

File: service/src/main/java/org/cbioportal/service/impl/ExpressionEnrichmentServiceImpl.java [hunk 5]
```diff
@@ -102,25 +109,50 @@
...
+            }
+            // Build filteredMolecularProfileCaseSets
+            filteredMolecularProfileCaseSets = new HashMap<>();
+            for (Map.Entry<String, List<MolecularProfileCaseIdentifier>> pair : molecularProfileCaseSets.entrySet()) {
+                Set<Integer> patientSet = new HashSet<Integer>();
+                List<MolecularProfileCaseIdentifier> identifierListUniqueByPatientId = new ArrayList<>();
+                for (MolecularProfileCaseIdentifier caseIdentifier : pair.getValue()) {
+                    if (patientSet.contains(sampleIdToPatientIdMap.get(caseIdentifier.getCaseId()))) {
+                        continue;
+                    } else {
```
Comment (kalletlak) [inline service/src/main/java/org/cbioportal/service/impl/ExpressionEnrichmentServiceImpl.java:131]: may be 
```
if (!patientSet.contains(sampleIdToPatientIdMap.get(caseIdentifier.getCaseId()))) {
    identifierListUniqueByPatientId.add(caseIdentifier);
    patientSet.add(sampleIdToPatientIdMap.get(caseIdentifier.getCaseId()));
}
```

File: core/src/main/scripts/importer/validateData.py [hunk 6]
```diff
@@ -4316,6 +4357,22 @@ def checkConsistencyFeatures(self):
 
         super(MultipleDataFileValidator, self).onComplete()
 
+    def checkId(self):
```
[the remark this answers is not in the document]
Comment (dippindots) [reply core/src/main/scripts/importer/validateData.py:4360]: Sure, I can do that

**Review by kalletlak — APPROVED** [review]

**Comment by pieterlukasse** [comment]
@n1zea144 @kalletlak I think the questions @n1zea144 raised make sense, but I don't see an answer to the last ones. Can you please add the rationale for pushing forward with this change? Why not just upload the same data through the existing sample level format, if it ends up in this way in the DB anyway?

**Pull request merged**